src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs

```diff
@@ -50,20 +58,33 @@ public RoamingVisualStudioProfileOptionPersister(IThreadingContext threadingCont
         {
             Contract.ThrowIfNull(globalOptionService);
 
-            _settingManager = (ISettingsManager)serviceProvider.GetService(typeof(SVsSettingsPersistenceManager));
             _globalOptionService = globalOptionService;
 
-            // While the settings persistence service should be available in all SKUs it is possible an ISO shell author has undefined the
-            // contributing package. In that case persistence of settings won't work (we don't bother with a backup solution for persistence
-            // as the scenario seems exceedingly unlikely), but we shouldn't crash the IDE.
-            if (_settingManager != null)
-            {
-                var settingsSubset = _settingManager.GetSubset("*");
-                settingsSubset.SettingChangedAsync += OnSettingChangedAsync;
-            }
+            var serviceContainer = serviceProvider.GetService<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+            _serviceBroker = serviceContainer.GetFullAccessServiceBroker();
+            _serviceBroker.AvailabilityChanged += Sb_AvailabilityChanged;
+
+            ConnectSettingsManager();
+        }
+
+        private void ConnectSettingsManager()
+        {
+            ThreadingContext.JoinableTaskFactory.Run(
+                async () =>
+                {
+                    _brokeredSettingsManager = await _serviceBroker.GetProxyAsync<IBrokeredSettingsManager>(VisualStudioServices.VS2019_4.SettingsManager);
+                    await _brokeredSettingsManager.RequestChangeEventsAsync("*", CancellationToken.None).ConfigureAwait(true);
+                    _brokeredSettingsManager.SettingChanged += OnSettingChanged;
+                    return;
+                });
+        }
+
+        private void Sb_AvailabilityChanged(object sender, ServiceHub.Framework.BrokeredServicesChangedEventArgs e)
+        {
+            ConnectSettingsManager();
         }
 
-        private System.Threading.Tasks.Task OnSettingChangedAsync(object sender, PropertyChangedEventArgs args)
+        private void OnSettingChanged(object sender, PropertyChangedEventArgs args)
```

Comment:
@JoshuaBStevens 

If I understand this code right, we get a notification about each property that changed and then we call the service back to fetch the current value of that property. Is that correct?
If so, then I don't think this is the best approach for a remote service. It would be better if the event carried the new value of the property along with the property name, so that we don't need to call back to fetch the value.
Also it should batch all the properties that changed, not send one at a time.

Reply:
Good suggestions.  I've opened DevDiv bug 1051672 to track this.  From your point of view, how do you imagine the batching working?  Implicit based on a timeout (e.g. as long as there are changes happening within 100ms of one another they will all be batched) or explicit based on e.g. an IDisposable's scope?